### PR TITLE
Add Azure Key Vault secrets provider

### PR DESCRIPTION
**Pull request by JonathanNathanson**
I'll open this with saying this is the first time I've ever contributed to an open source project, including this Nautobot app. So, if I've done anything wrong, apologies in advance and just let me know how do it correctly in future.

I have a need to use Azure Key Vault as a secrets provider in my own deployment, and given the existing issue on the matter (#20) hasn't received any traction for a while, I figured I'd have a go at building it myself. 

I haven't built any tests for the code as yet, but open to looking into that as a next step.

### CODE REVIEW

**Comment by gsnider2195** [comment]
Thank you for the submission! This looks great so far, but it will take us a bit to get an environment set up to test it locally and give you feedback. One question I have is why the minimum Nautobot version was bumped up to `^2.1.5`?

**Comment by JonathanNathanson** [comment]
> One question I have is why the minimum Nautobot version was bumped up to `^2.1.5`?

That's a mistake from early on when I was figuring out the development environment, should have been discarded before my commit. I've pushed another to revert that back.

**Comment by DistantVoyager** [comment]
@glennmatthews ready for review. This is just to merge the forked branch into a new working branch https://github.com/nautobot/nautobot-app-secrets-providers/tree/azure-provider-branch where I'll make some minor edits and add additional docs

File: nautobot_secrets_providers/providers/azure.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+"""Secrets Provider for Azure Key Vault."""
+
+import json
+
+try:
+  from azure.identity import DefaultAzureCredential
+  from azure.keyvault.secrets import SecretClient
+  azure_available = True
+except (ImportError, ModuleNotFoundError):
```
Comment (HanlinMiao) [inline nautobot_secrets_providers/providers/azure.py:9]: This should be sufficient since `ModuleNotFoundError` is a subclass of `ImportError`
```suggestion
except ImportError:
```

File: nautobot_secrets_providers/providers/azure.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        """Required parameters for Azure Key Vault."""
+
+        vault_url = forms.CharField(
+            required=True,
+            help_text="The URL of the Azure Key Vault",
+        )
+        secret_name = forms.CharField(
+            required=True,
+            help_text="The name of the secret in the Azure Key Vault",
+        )
```
Comment (HanlinMiao) [inline nautobot_secrets_providers/providers/azure.py:36]: Sorry I am not super familiar with the secrets provider, how/where are we using the `ParametersForm`?

Comment (DistantVoyager) [reply nautobot_secrets_providers/providers/azure.py:36]: It's for this Parameters form section when adding a new secret. It updates the form fields dynamically based on what provider you select. Believe it relates to this code in core: https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/secrets/providers.py 
image

File: nautobot_secrets_providers/providers/azure.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        # Assuming the secret value is JSON, parse it.
+        try:
+            data = json.loads(secret_value)
+        except json.JSONDecodeError as err:
+            # If JSON parsing fails, the secret might just be a string.
+            # In that case, just return the secret value directly.
+            return secret_value
+
+        # If it's a JSON object, return the entire JSON object.
+        return data
```
Comment (glennmatthews) [inline nautobot_secrets_providers/providers/azure.py:70]: Are JSON-structured secrets common in Azure? I'm a bit concerned about potentially returning a dict or other complex object here when pretty much all of the secrets logic in core assumes it'll be a string, I believe?

File: pyproject.toml [hunk 4]
```diff
@@ -31,6 +31,8 @@ hvac = { version = ">=0.11.0, <1.1.0", optional = true }
 nautobot = "^2.0.0"
 python = ">=3.8,<3.12"
 python-tss-sdk = {version = "~1.2.0", optional = true}
+azure-identity = "^1.15.0"
+azure-keyvault-secrets = "^4.8.0"
```
Comment (glennmatthews) [inline pyproject.toml:35]: These should be added as `optional = true` like the other provider-specific dependencies.

File: nautobot_secrets_providers/providers/azure.py (hunk 3 above)
Comment (DistantVoyager) [reply nautobot_secrets_providers/providers/azure.py:70]: I'd think they're fairly uncommon and can probably just remove the json parsing, unless @JonathanNathanson there was some particularity you had in mind when creating this

Comment (JonathanNathanson) [reply nautobot_secrets_providers/providers/azure.py:70]: Ah, I believe this was some logic from before I limited the scope to just simple secrets. By way of explanation, Azure Key Vault stores connection strings (eg for a database) as JSON objects - connection strings are  different type of secret not in scope for Nautobot secrets providers, so I wouldn’t expect any of the secrets in the case of this provider to be JSON. 

TLDR; the JSON parsing can be removed.
